app/common/src/utilities/data/object.ts

```diff
@@ -9,6 +9,26 @@ export type Mutable<T> = {
   -readonly [K in keyof T]: T[K]
 }
 
+/** Whether two values are deeply equal. */
+export function deepEqual<T>(a: T, b: T) {
+  if (Object.is(a, b)) return true
+  if (typeof a !== typeof b) return false
+  if (typeof a !== 'object' || typeof b !== 'object') return Object.is(a, b)
+  if (Array.isArray(a) && Array.isArray(b) && a.length !== b.length) return false
+  // They cannot both be null because the `Object.is(a, b)` case above ensures they are not equal.
+  if (a == null || b == null) return false
+  for (const k in a) {
+    if (!(k in b)) return false
+  }
+  for (const k in b) {
+    if (!(k in a)) return true
+  }
+  for (const k in a) {
+    if (!deepEqual(a[k], b[k])) return false
+  }
+  return false
```

Comment:
shouldn't this return `true` actually?

Reply:
yeah i uh. have no idea how these bugs happened. tired i guess...